### PR TITLE
Windows specific kubelet flags in kubeadm-flags.env

### PR DESCRIPTION
**Pull request by gab-satchi**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
kubeadm writes an environment file that's Linux specific. This PR removes the Linux specific flags and adds Windows specific values when kubeadm is being run on a Windows node. 

What is currently populated in a Windows node:
```
KUBELET_KUBEADM_ARGS="--cgroup-driver= --network-plugin=cni --pod-infra-container-image=k8s.gcr.io/pause:3.1"
```

What would be populated with the changes: 
```
KUBELET_KUBEADM_ARGS="--network-plugin=cni --pod-infra-container-image=mcr.microsoft.com/k8s/core/pause:1.2.0"
```

**Special notes for your reviewer**:
The concept of ENV files doesn't exist for native Windows services. The way to leverage the changes in this PR would be to use something like [nssm](https://nssm.cc/) to run a script that will execute the kubelet after reading the env file. With this approach, it will enable cluster api to override kubelet flags via metadata on Windows nodes. 

**Does this PR introduce a user-facing change?**:
```release-note
kubeadm: support Windows specific kubelet flags in kubeadm-flags.env
```

### CODE REVIEW

File: cmd/kubeadm/app/constants/constants_windows.go
Comment (neolit123) [inline cmd/kubeadm/app/constants/constants_windows.go]: i was under the impression that the long term goal is to make the pause image at GCR a multi-OS image and make it versioned as e.g. `3.2` to simplify the UX.
i'm hesitant to apply this change until this topic is clear.

this issue did not a give an answer to that:
https://github.com/kubernetes/kubernetes/issues/74218

also, the kubelet has also previously rejected defaulting differently between OSes.

File: cmd/kubeadm/app/constants/constants_windows.go
Comment (neolit123) [inline cmd/kubeadm/app/constants/constants_windows.go]: i was under the impression that the long term goal is to make the pause image at GCR a multi-OS image and make it versioned as e.g. `3.2` to simplify the UX.
i'm hesitant to apply this change until this topic is clear.

this issue did not a give an answer to that:
https://github.com/kubernetes/kubernetes/issues/74218

also, the kubelet has previously rejected defaulting differently between OSes (for flags at least).

File: cmd/kubeadm/app/images/images_windows.go
Comment (neolit123) [inline cmd/kubeadm/app/images/images_windows.go]: not sure if someone needs this at this point, but long term this hardcoding prevents customizing the image location for pause prepull.

File: cmd/kubeadm/app/phases/kubelet/flags_unix.go
Comment (neolit123) [inline cmd/kubeadm/app/phases/kubelet/flags_unix.go]: do you happen to know what is the state of ipv6 and dual-stack currently on Windows and if it is disabled, what are the plans for the near future?

i think this feature gate can be passed on Windows too, even if it's a NO-OP.

File: cmd/kubeadm/app/phases/kubelet/flags_windows.go
Comment (neolit123) [inline cmd/kubeadm/app/phases/kubelet/flags_windows.go]: overall i'm in favor of this change in particular.

but it makes sense to do the following:
- have one big buildKubeletArgMapCommon() function with a common setup:
  - hostname override setup
  - ipv6dualstack FG setup?
  - taints setup
- have abstracted buildKubeletArgMap() function on Windows that has:
  - CRI setup
- have abstracted buildKubeletArgMap() function on Linux that has:
  - CRI setup
  - resolve conf setup

**Comment by neolit123** [comment]
> Does this PR introduce a user-facing change?:
> NONE

please add a release note in the lines of:
> kubeadm: support Windows specific kubelet flags in kubeadm-flags.env

File: cmd/kubeadm/app/phases/kubelet/flags_windows.go
Comment (rosti) [reply cmd/kubeadm/app/phases/kubelet/flags_windows.go]: `resolv.conf` and `--pod-infra-container-image` can be counted as CRI setup (more accurately - Docker backend setup).
Ideally we should have these isolated somewhere even on Linux, as they are unlikely to get into the Kubelet component config. If the Docker shim moves, they are going to move with it.
Then the only thing left to do is the CRI. This, however, still does not imply having separate `buildKubeletArgMap` versions for Windows and Linux.

**Comment by gab-satchi** [comment]
Thanks for the comments @neolit123. Will update

**Review by rosti — COMMENTED** [review]
Thanks for taking on this @gab-satchi !
@neolit123 the big question here is, do we use the Docker shim on Windows? Is `--pod-infra-container-image` a valid option there?
If it is, then we can add more Windows specific code about fetching the correct pause image. If it isn't, then we can just `if runtime.GOOS != "windows" { ... }` all pause image stuff.

**Comment by neolit123** [comment]
@rosti 
> @neolit123 the big question here is, do we use the Docker shim on Windows? Is --pod-infra-container-image a valid option there?

docker shim and related flags are used on Windows.

File: cmd/kubeadm/app/phases/kubelet/flags_unix.go
Comment (gab-satchi) [reply cmd/kubeadm/app/phases/kubelet/flags_unix.go]: There was someone from MS that spoke about working on dual-stack support in our last sig call. Since that seems to be on the horizon, I'll pass this on to Windows as well.

File: cmd/kubeadm/app/images/images_windows.go
Comment (rhockenbury) [inline cmd/kubeadm/app/images/images_windows.go]: Latest windows pause image for 1903/1909 support is `mcr.microsoft.com/oss/kubernetes/pause:1.3.0`
 See https://github.com/kubernetes/kubernetes/issues/87339#issuecomment-582136209

File: cmd/kubeadm/app/phases/kubelet/flags_windows.go
Comment (neolit123) [reply cmd/kubeadm/app/phases/kubelet/flags_windows.go]: we had a discussion with @rosti about this during the kubeadm office hours and my proposal above to split the functions seems valid.

later on we want to move flags that are already in the v1beta1 KubeletConfiguration to the config.

File: cmd/kubeadm/app/phases/kubelet/flags_unix.go
Comment (neolit123) [reply cmd/kubeadm/app/phases/kubelet/flags_unix.go]: ok :+1:

File: cmd/kubeadm/app/images/images_windows.go
Comment (neolit123) [reply cmd/kubeadm/app/images/images_windows.go]: this can be formed like so:
```
func GetPauseImage(cfg *kubeadmapi.ClusterConfiguration) string {
	return GetGenericImage(
		"mcr.microsoft.com/k8s/core"
		"pause",
		"1.3.0"
	)
}
```

we discussed that hardcoding the image repository will be needed on windows for now.

File: cmd/kubeadm/app/images/images_others.go
Comment (neolit123) [inline cmd/kubeadm/app/images/images_others.go]: let's keep this file named `images_unix.go`, to match the pattern of `constants_*.go`

File: cmd/kubeadm/app/images/images_windows.go
Comment (neolit123) [inline cmd/kubeadm/app/images/images_windows.go]: i think it still makes sense to have PauseVersion in both constants_unix and constants_windows (with the different value).
also please have the commits squashed to 1.

**Comment by neolit123** [comment]
thanks for the updates.
this seems fine to me at this point.
leaving LGTM to @rosti

/approve
/priority backlog

File: cmd/kubeadm/app/images/images_windows.go
Comment (rhockenbury) [reply cmd/kubeadm/app/images/images_windows.go]: Please also note the path change `mcr.microsoft.com/oss/kubernetes` not `mcr.microsoft.com/k8s/core`

Comment (gab-satchi) [reply cmd/kubeadm/app/images/images_windows.go]: Thanks for the catch!

File: cmd/kubeadm/app/images/images_windows.go
Comment (rosti) [inline cmd/kubeadm/app/images/images_windows.go]: This is likely to cause some problems with air gaped environments. Suppose the following:
- Users are in an air gaped environment
- Users have a local container registry with all Windows images in it.
- They have overwritten the kubeadm config to use that container registry.
In that case we are going to pull everything from the local registry except the Windows `pause` image which is going to be hardwired to `mcr.microsoft.com/oss/kubernetes` and will require internet access or manual intervention on each Windows node.

The right thing to do here is to have pause image as a fat multi-platform image in the GCR repo and not have separate version of this function. Another possibility is to switch to `mcr.microsoft.com/oss/kubernetes` as the default image repository for Windows nodes. This would require the addition of kube-proxy to it (when kube-proxy starts to be executed as a DaemonSet).
A third possibility is to have an ugly hack in kubeadm, that is going to check if the image repository is the default gcr one and replace it with `mcr.microsoft.com` only in the case of `pause`.

@neolit123 WDYT?

File: cmd/kubeadm/app/phases/kubelet/flags_windows.go
Comment (rosti) [reply cmd/kubeadm/app/phases/kubelet/flags_windows.go]: The Windows version of `buildKubeletArgMap` is now a subset of the Linux one. How about we move it into a common space and name it `buildKubeletArgMapCommon`. Then have the Linux version of `buildKubeletArgMap` call it and add the rest of the flags. The Windows version of `buildKubeletArgMap` would be just a one liner (call `buildKubeletArgMapCommon`).

@gab-satchi @neolit123 WDYT?

File: cmd/kubeadm/app/images/images_windows.go
Comment (neolit123) [reply cmd/kubeadm/app/images/images_windows.go]: i forgot about air gapped use case and given the complications around multi-OS manifests and licensing, i think we might want to go with the hack solution to solve the problem for `kubeadm config images` too.

File: cmd/kubeadm/app/phases/kubelet/flags_windows.go
Comment (neolit123) [reply cmd/kubeadm/app/phases/kubelet/flags_windows.go]:
```
	if opts.nodeRegOpts.CRISocket == constants.DefaultDockerCRISocket {
		// These flags should only be set when running docker
		kubeletFlags["network-plugin"] = "cni"
		if opts.pauseImage != "" {
			kubeletFlags["pod-infra-container-image"] = opts.pauseImage
		}
	} else {
		kubeletFlags["container-runtime"] = "remote"
		kubeletFlags["container-runtime-endpoint"] = opts.nodeRegOpts.CRISocket
	}
```

can be moved to common.
then Linux has to have the explicit:

```
	if opts.nodeRegOpts.CRISocket == constants.DefaultDockerCRISocket {
		driver, err := kubeadmutil.GetCgroupDriverDocker(opts.execer)
		if err != nil {
			klog.Warningf("cannot automatically assign a '--cgroup-driver' value when starting the Kubelet: %v\n", err)
		} else {
			kubeletFlags["cgroup-driver"] = driver
		}
	}
```

and Windows can only call buildKubeletArgMapCommon()

File: cmd/kubeadm/app/images/images_windows.go
Comment (rosti) [reply cmd/kubeadm/app/images/images_windows.go]: Another possibility is to stop managing `pause` from kubeadm. It's really a CRI thing and we can add it as a requirement for folks to pre-pull manually after they install Docker/containerd/CRI-O.
That way we can get the same experience for all container runtimes.

Comment (neolit123) [reply cmd/kubeadm/app/images/images_windows.go]: we'd need a GA deprecation announcement of sorts to allow smoother transition for that.
it will certainly break the existing user base that assumes kubeadm will pull pause for them on new nodes.

**Comment by neolit123** [comment]
@gab-satchi 
would you be able to update based on the comments here:
https://github.com/kubernetes/kubernetes/pull/88287#discussion_r383360915
https://github.com/kubernetes/kubernetes/pull/88287#discussion_r383367299

next week is code-free, so let me know if you don't have the time and i can finish these items.

**Comment by gab-satchi** [comment]
@neolit123 I'm back today so I can take a look and update.

**Comment by neolit123** [comment]
@gab-satchi  thanks.

> next week, Friday is code-freeze, 

correction, it's Thursday this week.

File: cmd/kubeadm/app/images/images_unix.go
Comment (neolit123) [inline cmd/kubeadm/app/images/images_unix.go]: 2020

File: cmd/kubeadm/app/images/images_windows.go
Comment (neolit123) [inline cmd/kubeadm/app/images/images_windows.go]: 2020

File: cmd/kubeadm/app/phases/kubelet/flags_unix.go
Comment (neolit123) [inline cmd/kubeadm/app/phases/kubelet/flags_unix.go]: 2020

File: cmd/kubeadm/app/phases/kubelet/flags_windows.go
Comment (neolit123) [inline cmd/kubeadm/app/phases/kubelet/flags_windows.go]: 2020

File: cmd/kubeadm/app/images/images_windows.go
Comment (neolit123) [inline cmd/kubeadm/app/images/images_windows.go]: this empty line can be omitted.

File: cmd/kubeadm/app/images/images_windows.go
Comment (neolit123) [inline cmd/kubeadm/app/images/images_windows.go]: non-blocking for the PR:
@rosti didn't we decide to use `kubeadmapiv1` for import aliases everywhere?

**Comment by neolit123** [comment]
@gab-satchi thanks for the update
added minor comments to update the copyright dates.

**Comment by neolit123** [comment]
also please keep the commits squashed to 1.

File: cmd/kubeadm/app/constants/constants_unix.go
Comment (neolit123) [inline cmd/kubeadm/app/constants/constants_unix.go]: sorry if wasn't clear, the copyright for old files should not change. we should use 2020 for newly created ones.

File: cmd/kubeadm/app/constants/constants_windows.go
Comment (neolit123) [inline cmd/kubeadm/app/constants/constants_windows.go]: here as well.

Comment (gab-satchi) [reply cmd/kubeadm/app/constants/constants_windows.go]: whoops sorry. Somehow thought I split those up in this PR.

**Comment by neolit123** [comment]
waiting on the verify jobs to finish but this LGTM.
thanks

**Review by rosti — APPROVED** [review]
Thanks @gab-satchi !
I think the unit test situation with `buildKubeletArgMapCommon` can be better. However, I think it's OK for now.
/lgtm
/hold cancel

**Pull request merged**
